File: Problemset/1624-Largest Substring Between Two Equal Characters/Solution.rb
# @param {String} s
# @return {Integer}
def max_length_between_equal_characters(s)
  pairs = Array.new(26) { |_| [s.length, 0] }

  s.chars.each_with_index do |c, i|
    k = c.ord - 97

    pairs[k][0] = i if pairs[k][0] > i
    pairs[k][1] = i if pairs[k][1] < i
  end

  pairs.map { |pair| pair[1] - pair[0] - 1 }.max
end

File: Problemset/1624-Largest Substring Between Two Equal Characters/Solution.rs
impl Solution {
    pub fn max_length_between_equal_characters(s: String) -> i32 {
        let mut pairs = vec![(s.len() as i32, 0); 26];

        for (i, c) in s.bytes().enumerate() {
            let pair = &mut pairs[(c - b'a') as usize];

            pair.0 = pair.0.min(i as i32);
            pair.1 = pair.1.max(i as i32);
        }

        pairs.iter().map(|(i, j)| j - i - 1).max().unwrap()
    }
}
